pennylane/transforms/decompositions/clifford_t_transform.py

```diff
@@ -330,12 +331,71 @@ def _merge_param_gates(operations, merge_ops=None):
     return merged_ops, number_ops
 
 
-# pylint: disable=too-many-branches
+@lru_cache(maxsize=2000)
+def _map_wires(op, wire):
+    """Maps the operator to the provided wire."""
+    return op.map_wires({0: wire})
+
+
+class _CachedCallable:
+    """A class to cache the decomposition of the operators.
+
+    Args:
+        method (str): The method to be used for decomposition.
+        epsilon (float): The maximum permissible operator norm error for the decomposition.
+        cache_size (int): The size of the cache built for the decomposition function based on the angle.
+        **method_kwargs: Keyword argument to pass options for the ``method`` used for decompositions.
+    """
+
+    def __init__(self, method, epsilon, cache_size, **method_kwargs):
+        match method:
+            case "sk":
+                self.decompose_fn = lru_cache(maxsize=cache_size)(
+                    partial(sk_decomposition, epsilon=epsilon, **method_kwargs)
+                )
+            case _:
+                raise NotImplementedError(
+                    f"Currently we only support Solovay-Kitaev ('sk') decomposition, got {method}"
+                )
+
+        self.method = method
+        self.epsilon = epsilon
+        self.cache_size = cache_size
+        self.method_kwargs = method_kwargs
+        self.query = lru_cache(maxsize=cache_size)(self.cached_decompose)
+
+    def compatible(self, method, epsilon, cache_size, **method_kwargs):
+        """Check equality based on `method`, `epsilon` and `method_kwargs`."""
+        return (
+            self.method == method
+            and self.epsilon <= epsilon
+            and self.cache_size <= cache_size
+            and self.method_kwargs == method_kwargs
+        )
+
+    def cached_decompose(self, op):
+        """Decomposes the angle into a sequence of gates."""
+        f_adj = op.data[0] < 2 * math.pi  # 4 * pi / 2
+        cached_op = op if f_adj else op.adjoint()
+
+        seq = self.decompose_fn(cached_op)
+        if f_adj:
+            return seq
+
+        adj = [qml.adjoint(s, lazy=False) for s in reversed(seq)]
+        return adj[1:] + adj[:1]
+
+
+_CLIFFORD_T_CACHE = None
```

Comment:
Could you add this to the top of the file and add a code comment describing the purpose?